### PR TITLE
refactor: memoize fetchErc20Decimals, relocate to utils/token, and apply to other instances

### LINKED ISSUE #24374: fetchErc20Decimals should not use getTokenStandardAndDetails
We are making 3-4 additional RPC requests for each ERC20 token during simulation renders. Given that 90% of transactions involve only a single token, and many of these are not ERC20 tokens, this extra load is generally manageable. However, for the 10% of transactions involving multiple ERC20 tokens, this can quickly add up.

For example, transactions with 10 or more tokens could result in 20-30 additional requests, which might impact performance and rate limits.

@bergeron [wrote](https://consensys.slack.com/archives/GTQAGKY5V/p1714773613632259?thread_ts=1714771396.130019&cid=GTQAGKY5V):

> I do suspect getTokenStandardAndDetails is used in some places it shouldn't.  It should only be used when we have an address and no idea what kind of token it is.  Because it has to cycle through all the standards.  For example this code just wants ERC20 decimals:
> ```
> async function fetchErc20Decimals(address: Hex): Promise<number> {
>   try {
>     const { decimals } = await getTokenStandardAndDetails(address);
>     return decimals ? parseInt(decimals, 10) : ERC20_DEFAULT_DECIMALS;
>   } catch {
>     return ERC20_DEFAULT_DECIMALS;
>   }
> }
> ```
> Which should be 1 RPC request since we already 'know' its an ERC20, but that could do up to 4 RPC requests (checking for NFTs and returning other ERC20 details other than decimals)

### PR DESCRIPTION
**Pull request by digiwand**
## **Description**

- move fetchErc20Decimals method to utils/token 
- use fetchErc20Decimals for other instances calling getTokenStandardAndDetails. This will set decimals to 18 if no decimals were retrieved in the response. Previously, it used 0 instead of 18 in these cases.
- memoize method to avoid making additional requests for the same token - performance improvement
- create tests for memorized fetchErc20Decimals method 

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27088?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/24374

## **Manual testing steps**

1. Test token values and fiat conversions on confirmation pages

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/pages/confirmations/constants/token.ts
Comment (matthewwalsh0) [inline ui/pages/confirmations/constants/token.ts:1]: Minor, given there is very little in `index.ts` currently, do we warrant this level of modularity?

**Review by matthewwalsh0 — DISMISSED** [review]

File: ui/pages/confirmations/constants/token.ts
Comment (digiwand) [reply ui/pages/confirmations/constants/token.ts:1]: questionable, so updated this to share `ui/pages/confirmations/utils/token.ts`

**Review by matthewwalsh0 — DISMISSED** [review]

**Comment by digiwand** [comment]
oops, accidentally merged a commit. Reverted after. May I get a reapproval please? @matthewwalsh0

**Review by jpuri — APPROVED** [review]
👍

**Review by matthewwalsh0 — APPROVED** [review]

**Pull request merged**